### PR TITLE
feat(middleware): issues warnings when using node.js global APIs in middleware

### PR DESCRIPTION
**Pull request by feugy**
## What's in there?

Middleware, since they are ran on edge runtime, do not have access to Node.js globals. While #36434 addressed cases where a middleware would import a node.js module, this PR issues warning when trying to access unsupported globals, such as:
`clearImmediate()`, `process.cwd()`, `Event` or `TextDecoderStream` (full list in `packages/next/shared/lib/constants.ts`).

The new behavior is:
- a warning at build time. Failing the build would be too strict: it's possible for the final bundle to include unsupported calls that would never be reached at run time (despite tree-shaking).
- a warning on the first access at run time (useful in development)
- `process.env` is always allowed
- other attributes/methods of `process` are not supported, but could be overridden

## How to test?

First, build next: `yarn workspace next release`

Then, using integration tests:
1. `yarn jest test/integration/middleware-with-node.js-apis` (checks both build and run time behavior)
1. `yarn jest test/integration/middleware-overrides-node.js-api` (ability to override `process.XYZ` and use `process.env` with no warning)

Finally, as an example:
1. create a middleware file in `examples/with-node-api/middleware.js`, with this content:
   ```js
   import { NextResponse } from 'next/server'
   
   export default function middleware() {
     console.log(`this should work:`, process.env, __dirname, __filename)
     console.log(`this should not`, setImmediate(() => {}))
     return NextResponse.next()
   }
   ```
1. creates a plain pages in  `examples/with-node-api/pages/index.js`, with this content:
   ```js
   export default function Home() {
     return <div>A page</div>
   }
   ```
3. run it in dev mode  `yarn next dev examples/with-node-api/` and browse to https://localhost:3000:
   > it should print `this should work: { NEXT_RUNTIME: 'edge' }`, issue a warning about using `setImmediate()`, and fail because `setImmediate()` is not detined.
4. build it: `yarn next build examples/with-node-api/`:
   > it should build, and issue a warning about using `setImmediate()`,

## Notes to reviewers

- `build/webpack/plugins/middleware-plugin.ts` does the build-time validation.
- `server/web/sandbox/context.ts` does the run-time behavior.

To determine unsupported APIs, I compared the list from [node.js docs](https://nodejs.org/docs/latest-v17.x/api/globals.html) to the our [edge-runtime](https://nextjs.org/docs/api-reference/edge-runtime).
Please note that `__dirname` and `__filename` are still allowed because webpack uses them.

## Feature

- [ ] Implements an existing feature request or RFC. Make sure the feature request has been accepted for implementation before opening a PR.
- [x] Related issues linked using `fixes #number`
- [x] Integration tests added
- [ ] Documentation added
- [ ] Telemetry added. In case of a feature if it's used or not.
- [x] Errors have helpful link attached, see `contributing.md`

## Documentation / Examples

- [x] Make sure the linting passes by running `yarn lint`

### CODE REVIEW

File: packages/next/shared/lib/constants.ts [hunk 1]
```diff
@@ -67,3 +67,33 @@ export const OPTIMIZED_FONT_PROVIDERS = [
 ]
 export const STATIC_STATUS_PAGES = ['/500']
 export const TRACE_OUTPUT_VERSION = 1
+
+export const EDGE_UNSUPPORTED_NODE_APIS = [
+  'clearImmediate',
+  'setImmediate',
+  'structuredClone',
+  'queueMicrotask', // TODO allow that one?
```
Comment (Schniz) [inline packages/next/shared/lib/constants.ts:75]: @Kikobeats wdyt

File: test/integration/middleware/with-node.js-apis/test/index.test.ts [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+  waitFor,
+} from 'next-test-utils'
+import { join } from 'path'
+
+jest.setTimeout(1000 * 60 * 2)
+
+const unsupportedFunctions = [
+  'setImmediate',
+  'clearImmediate',
+  'structuredClone',
```
Comment (nuta) [inline test/integration/middleware/with-node.js-apis/test/index.test.ts:18]: I think `structuredClone` is supported in the edge runtime.

File: packages/next/shared/lib/constants.ts [hunk 3]
```diff
@@ -67,3 +67,33 @@ export const OPTIMIZED_FONT_PROVIDERS = [
 ]
 export const STATIC_STATUS_PAGES = ['/500']
 export const TRACE_OUTPUT_VERSION = 1
+
+export const EDGE_UNSUPPORTED_NODE_APIS = [
+  'clearImmediate',
+  'setImmediate',
+  'structuredClone',
```
Comment (nuta) [inline packages/next/shared/lib/constants.ts:74]: ditto

File: packages/next/shared/lib/constants.ts (hunk 1 above)
Comment (feugy) [reply packages/next/shared/lib/constants.ts]: allowed, thanks!

File: test/integration/middleware/with-node.js-apis/test/index.test.ts (hunk 2 above)
Comment (feugy) [reply test/integration/middleware/with-node.js-apis/test/index.test.ts]: allowed, thanks!

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 9e2a6d10eee2649087fde28154beefd4617836f4

`yarn testheadless test/integration/cli/test/index.test.js` 
- CLI Usage > info > should print output

<details>
<summary>Expand output</summary>

  ● CLI Usage › info › should print output

    expect(received).toBe(expected) // Object.is equality

    Expected: ""
    Received: "warn  - Latest canary version not detected, detected: \"12.1.7-canary.10\", newest: \"12.1.7-canary.9\".
            Please try the latest canary version (`npm install next@canary`) to confirm the issue still exists before creating a new issue.
            Read more - https://nextjs.org/docs/messages/opening-an-issue
    "

      496 |       // warning will show so skip this check for the stable release
      497 |       if (pkg.version.includes('-canary')) {
    > 498 |         expect(info.stderr || '').toBe('')
          |                                   ^
      499 |       }
      500 |       expect(info.stdout).toMatch(
      501 |         new RegExp(`

      at Object.<anonymous> (integration/cli/test/index.test.js:498:35)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Review by timneutkens — APPROVED** [review]

**Pull request merged**